src/Interpreters/Cache/FileCache.cpp

```diff
@@ -638,7 +638,7 @@ FileCache::getOrSet(
 
     assertInitialized();
 
-    FileSegment::Range initial_range(offset, offset + size - 1);
+    FileSegment::Range initial_range(offset, std::min(offset + size, file_size) - 1);
```

Comment:
This change does not fix the assert, but it fixes new unit test case added by me in this PR (there are several added tests, but the one that required this change is in line 1479 of gtest_filecache), in practice this is very unlikely to be hit.